core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
           versions().subList(versions().size() - versionHistorySizeToKeep, versions().size());
       List<ViewHistoryEntry> history =
           history().subList(history().size() - versionHistorySizeToKeep, history().size());
-      return ImmutableViewMetadata.builder().from(this).versions(versions).history(history).build();
+      return ViewMetadata.buildFrom(this)
+          .discardChanges()
+          .versions(versions)
+          .history(history)
+          .build();
     }
 
     return this;
   }
+
+  static Builder builder() {
+    return new Builder();
+  }
+
+  static Builder buildFrom(ViewMetadata base) {
+    return new Builder(base);
+  }
+
+  class Builder {
+    private int formatVersion = 1;
+    private String location;
+    private Integer currentSchemaId;
+    private List<Schema> schemas = Lists.newArrayList();
+    private int currentVersionId = 1;
+    private List<ViewVersion> versions = Lists.newArrayList();
+    private List<ViewHistoryEntry> history = Lists.newArrayList();
+    private Map<String, String> properties = Maps.newHashMap();
+    private List<MetadataUpdate> changes = Lists.newArrayList();
+
+    private Builder() {}
+
+    private Builder(ViewMetadata base) {
+      this.formatVersion = base.formatVersion();
+      this.location = base.location();
+      this.currentSchemaId = base.currentSchemaId();
+      this.schemas = Lists.newArrayList(base.schemas());
+      this.currentVersionId = base.currentVersionId();
+      this.versions = Lists.newArrayList(base.versions());
+      this.history = Lists.newArrayList(base.history());
+      this.properties = Maps.newHashMap(base.properties());
+      this.changes = Lists.newArrayList(base.changes());
+    }
+
+    public Builder formatVersion(int newFormatVersion) {
+      this.formatVersion = newFormatVersion;
+      this.changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      return this;
+    }
+
+    public Builder location(String newLocation) {
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder currentVersionId(int versionId) {
+      this.currentVersionId = versionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(versionId));
+      return this;
+    }
+
+    public Builder currentSchemaId(Integer schemaId) {
+      this.currentSchemaId = schemaId;
+      this.changes.add(new MetadataUpdate.SetCurrentSchema(schemaId));
+      return this;
+    }
+
+    public Builder addSchema(Schema schema) {
+      this.schemas.add(schema);
+      this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
+      return this;
+    }
+
+    public Builder addAllSchemas(Iterable<Schema> schemaEntries) {
+      for (Schema schema : schemaEntries) {
+        this.schemas.add(schema);
+        this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
```

Comment:
Do you think we should make this optional? I don't think that `highestFieldId` is actually the last assigned field ID in all cases so it would probably be better not to send it.

Reply:
I think it's a good idea to make `highestFieldId` optional here. Also looks like this field is already optional in the REST Spec [here](https://github.com/apache/iceberg/blob/621d301b17417af4b71a0e8c93db9db6f8266839/open-api/rest-catalog-open-api.yaml#L1677) but not in the code. I'll open a separate PR to make this optional in the code as well, so that we don't have to set/send it for views

Reply:
looks like we default to `schema.highestFieldId()` when this `last-column-id` isn't send: https://github.com/apache/iceberg/blob/26af4c7387d68a04762bb50d388d3227536527b9/core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java#L404

Reply:
I looked into making this optional by moving `AddSchema.lastColumnId` from an  `int` to an `Integer` and not passing/sending it for views, but that is an API-breaking change unfortunately. I think we're probably ok sending `new MetadataUpdate.AddSchema(schema, schema.highestFieldId())` for Views, since that's what https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java#L404 would be doing anyway

Reply:
If we aren't going to make it optional, then we need to send `max(_.highestFieldId()` for all known schemas.

Reply:
should be fixed now